client-tools/Shell/ProcessMonitoringFeature.cpp

```diff
@@ -0,0 +1,147 @@
+
+#include "ProcessMonitoringFeature.h"
+
+using namespace arangodb::basics;
+using namespace arangodb::options;
+// using namespace arangodb::rest;
+using namespace arangodb;
+
+namespace arangodb {
+
+void ProcessMonitoringFeature::addMonitorPID(ExternalId& pid) {
+  MUTEX_LOCKER(mutexLocker, _MonitoredExternalProcessesLock);
+  _monitoredProcesses.push_back(pid);
+}
+
+void ProcessMonitoringFeature::removeMonitorPID(ExternalId const& pid) {
+  MUTEX_LOCKER(mutexLocker, _MonitoredExternalProcessesLock);
+  for (auto it = _monitoredProcesses.begin(); it != _monitoredProcesses.end();
+       ++it) {
+    if (it->_pid == pid._pid) {
+      _monitoredProcesses.erase(it);
+      break;
+    }
+  }
+}
+
+std::optional<ExternalProcessStatus>
+ProcessMonitoringFeature::getHistoricStatus(TRI_pid_t pid) {
+  MUTEX_LOCKER(mutexLocker, _MonitoredExternalProcessesLock);
+  auto it = _ExitedExternalProcessStatus.find(pid);
+  if (it == _ExitedExternalProcessStatus.end()) {
+    return std::nullopt;
+  }
+  return std::optional<ExternalProcessStatus>{it->second};
+}
+
+ProcessMonitoringFeature::ProcessMonitoringFeature(Server& server)
+    : ArangoshFeature(server, *this) {
+  startsAfter<V8SecurityFeature>();
+}
+
+ProcessMonitoringFeature::~ProcessMonitoringFeature() = default;
+
+void ProcessMonitoringFeature::validateOptions(
+    std::shared_ptr<options::ProgramOptions>) {
+  _enabled =
+      server().getFeature<V8SecurityFeature>().isAllowedToControlProcesses();
+}
+
+void ProcessMonitoringFeature::start() {
+  if (_enabled) {
+    _monitorThread = std::make_unique<ProcessMonitorThread>(server(), this);
+    if (!_monitorThread->start()) {
+      LOG_TOPIC("33c33", FATAL, Logger::SYSCALL)
+          << "failed to launch monitoring background thread";
+      FATAL_ERROR_EXIT();
+    }
+  }
+}
+void ProcessMonitoringFeature::beginShutdown() {
+  if (_enabled) {
+    _monitorThread->shutdown();
+  }
+}
+
+void ProcessMonitorThread::run() {  // override
+  while (!isStopping()) {
+    try {
+      std::vector<ExternalId> mp;
+      {
+        MUTEX_LOCKER(mutexLocker,
+                     _processMonitorFeature->_MonitoredExternalProcessesLock);
+        mp = _processMonitorFeature->_monitoredProcesses;
+      }
+      for (auto const& pid : mp) {
+        auto status = TRI_CheckExternalProcess(pid, false, 0);
+        if ((status._status == TRI_EXT_TERMINATED) ||
+            (status._status == TRI_EXT_ABORTED) ||
+            (status._status == TRI_EXT_NOT_FOUND)) {
+          // Its dead and gone - good
+          _processMonitorFeature->removeMonitorPID(pid);
+          {
+            MUTEX_LOCKER(
+                mutexLocker,
+                _processMonitorFeature->_MonitoredExternalProcessesLock);
+            _processMonitorFeature->_ExitedExternalProcessStatus[pid._pid] =
+                status;
+          }
+          triggerV8DeadlineNow(false);
+        }
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment:
Wouldn't it be more efficient to let this sleep using a condition variable, and only signal it to wake up once something actually needs to be done?
Otherwise it will wake up 100 times a second, look a mutex, copy data etc.